# Assembling genomes

Right, so here we are! Let’s assemble the genome of *Apodemus sylvaticus*. You are going to run the assembler Hifiasm today.

Have a look at Hifiasm's website:

* [Hifiasm github repository](https://github.com/chhylp123/hifiasm)

Now let’s run the assembly on our subsamples.

First, make sure you are in the directory that you created during our first hands-on tutorial (`~/a_sylvaticus/`). Then let’s create a directory the hifiasm run.
```console  
mkdir hifiasm
```  
Make sure the conda environment is activated:

```console
conda activate BIO5025
```

Let’s change to the hifiasm folder and create a symlink to the subsample dataset of reads:

```console  
cd ~/a_sylvaticus/hifiasm/
ln -s path/mApoSyl1.60.HiFi.fasta
```

Now let’s call hifiasm and have a look at the parameters it prints:

```console  
hifiasm
```

And this is the line you will run:

```console  
hifiasm -f0 --primary -t 1 -o mApoSyl1.60.hifiasm mApoSyl1.60.HiFi.fasta
```

PS: the `-o` option defines the *prefix* which will be used to name all the output files generated by Hifiasm.

Let’s wait a few minutes for the assembler to run. It will be fast, but while it runs, you can take some time to search what the parameters we used for running hifiasm mean. 
You can do that by either looking at the website or by opening a third tab/window and running hifiasm on help mode (e.g. `hifiasm -h`).

## Interpreting our results

Hifiasm outputs different intermediate files together with the final assembly result. 

Among the results hifiasm produces, you will find the `<prefix>.p_ctg.gfa` file (remember that the prefix has been defined using the `-o` option when running Hifiasm, the file name is likely to be mApoSyl1.60.hifiasm.p_ctg.gfa). 
This is our assembly output file. Hifiasm gives us the [assembly graph](http://gfa-spec.github.io/GFA-spec/GFA1.html) as an output. 
This means you need to convert the graph into a fasta file. For that, we have an awk script named `gfa2fa`.

Add the directory containing that script to the PATH and run it:

```console 
export PATH=path:$PATH
gfa2fa <prefix>.p_ctg.gfa > <prefix>.p_ctg.fa
```

Right. So now, what to do with the assembly file?
Well, first you should have a look at the statistics for this assembly. Let’s run our asmstats script on it

```console  
asmstats <prefix>.p_ctg.fa > <prefix>.p_ctg.fa.stats
```

## WELL DONE!
You have just done eukaryotic genome assembly using PacBio HiFi reads and two different assemblers! **That is fantastic!**

Again, let's remember you have ran hifiasm only for a subset of reads, which means this is never going to get you a complete assembly. So, for now, keep this results you have and lets look at statistics for an assembly I ran previously with the complete reads dataset.

The files will be on the working data folder (`mApoSyl1_data/`) and will be named `mApoSyl1.hifiasm.total.[a/p]_ctg.fa.gz` (Hifiasm produces two different files, one with suffix `a_ctg.fa.gz` and another one with the suffix `p_ctg.fa.gz`). In your species home directory (`~/<species>/`), create a symlink for those files (with the `ln -s` command) and run `asmstats` on them (e.g. `asmstats <species_id>.hicanu.total.contigs.fasta.gz > <species_id>.hicanu.total.contigs.fasta.stats`). Then answer:
  
  1-) What is the assembled size, number of contigs and N50 for the hicanu assembly? 
  
  1a-) Is the Hicanu result close to the estimated genome size? Why?
  
  2-) Hifiasm generates two files. Why? What are the assembly metrics for these two files?
  
Now gather all these results, let’s go to the larger group and let’s discuss them together!













singularity exec --bind /lustre/:/lustre/ /software/tola/images/hifiasm-0.19.5.sif hifiasm -o hifiasm60again --primary -t 1 gbk.HiFiMapped.bam.filtered.60.fasta

blast: done

complete genome for asm stats

purging and merqury:

before perging:
file for asmstats, 
